File: backend-python/src/main/resources/org/shedlang/compiler/backends/python/modules/Stdlib/Platform/StringBuilder.py
from ...builtins import effect_handler_call, effect_handler_discard, effect_handler_push


class _StringBuilderEffect(object):
    _effect_id = -10
    _Exit = None


def build(func):
    string_builder = []

    def write(value):
        string_builder.append(value)

    effect_handler_push(_StringBuilderEffect, dict(
        write=write,
    ))

    func()

    effect_handler_discard()

    return "".join(string_builder)


def write(value):
    effect_handler_call(_StringBuilderEffect._effect_id, "write", value)

File: backend-javascript/src/main/resources/org/shedlang/compiler/backends/javascript/modules/Stdlib/Platform/StringBuilder.js
"use strict";

const stringBuilders = [];

function build(func) {
    stringBuilders.push([]);
    func();
    const stringBuilder = stringBuilders.pop();
    return stringBuilder.join("");
}

async function buildAsync(func) {
    stringBuilders.push([]);
    await func();
    const stringBuilder = stringBuilders.pop();
    return stringBuilder.join("");
}

build.async = buildAsync;

function write(value) {
    stringBuilders[stringBuilders.length - 1].push(value);
}

exports.build = build;
exports.write = write;
